packages/schematics/angular/utility/dependencies.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { JsonAstObject, JsonParseMode, parseJsonAst } from '@angular-devkit/core';
+import { SchematicsException, Tree } from '@angular-devkit/schematics';
+import { appendPropertyInAstObject, findPropertyInAstObject } from './json-utils';
+
+
+const pkgJsonPath = '/package.json';
+export enum NodeDependencyType {
+  Default = 'dependencies',
+  Dev = 'devDependencies',
+  Peer = 'peerDependencies',
+  Optional = 'optionalDependencies',
+}
+
+export interface NodeDependency {
+  type: NodeDependencyType;
+  name: string;
+  version: string;
+  overwrite?: boolean;
+}
+
+export function addPackageJsonDependency(tree: Tree, dependency: NodeDependency): void {
+  const packageJsonAst = _readPackageJson(tree);
+  const depsNode = findPropertyInAstObject(packageJsonAst, dependency.type);
+  const recorder = tree.beginUpdate(pkgJsonPath);
+
+  if (!depsNode) {
+    // Haven't found the dependencies key, add it to the root of the package.json.
+    appendPropertyInAstObject(recorder, packageJsonAst, dependency.type, {
+      [dependency.name]: dependency.version,
+    });
+  } else if (depsNode.kind === 'object') {
+    // check if package already added
+    const depNode = findPropertyInAstObject(depsNode, dependency.name);
+
+    if (!depNode) {
+      // Package not found, add it.
+      appendPropertyInAstObject(
+        recorder,
+        depsNode,
+        dependency.name,
+        dependency.version,
+      );
+    } else if (dependency.overwrite) {
+      // Package found, update version if overwrite.
+      const { end, start } = depNode;
+      recorder.remove(start.offset, end.offset - start.offset);
+      recorder.insertRight(start.offset, JSON.stringify(dependency.version));
+    }
+  }
+
+  tree.commitUpdate(recorder);
+  _sortDependencies(tree, dependency);
```

Comment:
Instead of doing this, why not read through the keys and find the first package name that `packageName.localeCompare(dependency.name) > 0` and insert before that key? `_sortDependencies` looks expensive and complex (reading, reparsing and rewriting the file for every key).

Reply:
You can call that method `insertPropertyInAstObjectInOrder` and add it to `json-utils`, what do you think?

Reply:
suggested that up [here](https://github.com/angular/devkit/pull/954#discussion_r190045801) 😄

Reply:
The issue is that this wouldn't clean up other unsorted packages in the object. Which is the behavior of the npm client.